### ISSUE TITLE
Eslint `no-floating-promises` false positive for PromiseLike?

### ISSUE THREAD

**Issue by ZiiMakc** [typescript-eslint/typescript-eslint#4961]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=4.6.4&sourceType=module&code=C4TwDgpgBAglC8UDeBYAUFKBjA9gEwgC4oAKASgQD5YBudTYACwgDtiSAzAVwBsOBLHjwh52FeNQBuOfngA0UAE4QAVhCzAR7CIsXEAorpyLxUmXlNRpsumgC+t9AEMAziBZYo3D8H44WUC7kyPTY-i7AUE7EcIgAjFEuUSwgjhhRAHS4BOS2mE4A7k78kU5Z+BC56HZAA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AQPZYBmuZAhhIQOZYwpkC2aCiIAXCHBRsUIAL5A&tsConfig=N4XyA

### Repro Code

```typescript
type A = {
  code: () => A;
  then: (fulfilled: () => void, rejected: (err: Error) => void) => void;
};

async function s() {
  const a: A = 1 as any;

  a.code(); // error @typescript-eslint/no-floating-promises
  await a.code(); // ok
}
```
```


### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-floating-promises": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {}
}
```

### Expected Result

Error should not be reported for `a.code()` without `await` as it does a mutation on  `PromiseLike` object and returns it.

### Actual Result

Error is reported for `a.code()`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.22.0` |
| `@typescript-eslint/parser`            | `5.22.0` |
| `TypeScript`                           | `4.6.4` |
| `ESLint`                               | `8.15.0` |
| `node`                                 | `16.15.0` |
